### PR TITLE
Half tier mods in optimiser

### LINKED ISSUE #6469: Half tier stats not working as they should be
**What version of DIM are you using?**
Version 6.49.0.1000177 (beta), built on 1/27/2021, 1:02:30 AM (blue icon)

**What Browser and OS are you using?**
Firefox 84.0.2 (64-bit)

**Describe the bug and how to make it happen**
The new update states: Loadout Optimizer shows "half tiers" (0.5) to help you see where a 5-point mod might bump you up a tier. Note that this does not suggest we're displaying a new 0.1 precision for tiers - we are specifically rounding down to half-tiers now instead of full tiers.

The feature is nice, but there are some things wrong with it atm. First of all half stats should not count towards the Tier total since it is straight up inflating the tier numbers without actually being a T31 (picture example). 
For some reason the selection pane on the left doesn't work properly, for example the discipline tier on the left always drops back to 2.5 even though there are higher builds possible. If i select 10 on recovery for example it drops down to the minimum as well, because 9 is the highest possible i can reach with my armor. 
image

### PR DESCRIPTION
**Pull request by ryan-rushton**
Adding the ability to quick add half tier mods in the optimiser.

image
image

There are a few changes that went into making this,
1. The mod key business was removed from the mod picker and added to the reducer. This allows us to add mods from anywhere and when we have multiple of the same ones it's easy to filter them out and have a key for react.
2. Removed the locked mod type from adding new mods, this allows us to just send pluggable inventory items to the reducer, this is a bit nicer when dealing with sockets and plugs. We no longer need to construct the LockedMod as the reducer does it.
3. Created a selector for pulling half tier mods out of items. This involves finding a single item with a general mod socket, then from that getting all its reusable plugs and looking for the mods which add 5 to an armour stat. There is a potential for future error there if bungie adds other +5 mods to the general socket pool.
4. Tweaked a few styles as we have 4 buttons now and it was getting a little cramped where they were sitting.
5. I haven't handled the case where a user might already have a half tier mod for a stat which makes it a half tier. Currently it will try and add another half tier mod, in a perfect world it would swap it for a +10. I figure users can do this themselves.
6. I also rejigged how we calculate the tallest row in generated sets. I was having issues with my new button layout covering the next row. As a bonus we also now dont have the issue of the swap icon sitting over the border, and swapping phone from landscape to portrait or vice versa will also recalculate the height.

Closes #6469

### CODE REVIEW

File: src/app/dim-ui/RichDestinyText.tsx [hunk 1]
```diff
@@ -147,18 +147,13 @@ const generateConversionTable = _.once((defs: D2ManifestDefinitions) => {
...
   );
   return (
     (replacement && (
-      <img
-        src={replacement.icon}
-        className={styles.inlineSvg}
-        title={textSegment}
-        key={textSegment}
-      />
+      <img src={replacement.icon} className={styles.inlineSvg} title={textSegment} key={index} />
```
Comment (ryan-rushton) [inline src/app/dim-ui/RichDestinyText.tsx:156]: We actually have text that has two barrier icons in it. This was causing a react key error. I just swapped it to an index as its not important rendering wise and it removes the error.

File: src/app/loadout-builder/mod-utils.ts [hunk 2]
```diff
@@ -132,3 +132,14 @@ export function assignModsToArmorSet(
 
   return [assignedMods, unassignedMods];
 }
+
+/**
+ * Checks to see if some mod in a collection of LockedMod or LockedMod,
+ * has an elemental (non-Any) energy requirement
+ */
+export function someModHasEnergyRequirement(mods: LockedMod[]) {
```
Comment (ryan-rushton) [inline src/app/loadout-builder/mod-utils.ts:140]: This has just been moved as I ran into a circular dependency with it.

File: src/app/loadout-builder/generated-sets/GeneratedSets.tsx [hunk 3]
```diff
@@ -10,15 +11,64 @@ import {
...
+
+/**
+ * Gets the set used to measure how high a row can be. It also returns a recalcTrigger that is
+ * intended to trigger a recalculation when the value changes.
+ *
+ * It figures out the tallest row by looking at items with exotic perks and swap icons.
+ * Exotic perks add another row to the mod icons and the swap icon sits below the item image.
+ * The height they add is roughly equivalent so we treat both conditions equally.
+ */
+function getMeasureSetAndIndicator(sets, isPhonePortrait): [ArmorSet | undefined, number] {
```
Comment (ryan-rushton) [inline src/app/loadout-builder/generated-sets/GeneratedSets.tsx:36]: This fixes issues where we could see the swap icon falling below the bottom border. Also mobile rows are correctly sized now. I was having issues with my new buttons sitting over the next row. This is all fixed now

File: src/app/loadout-builder/generated-sets/GeneratedSetButtons.tsx [hunk 4]
```diff
@@ -40,19 +48,73 @@ export default function GeneratedSetButtons({
...
+    if (value % 10 > 4) {
+      statsWithPlus5.push(statHashes[stat]);
+    }
+  }
+
+  /**
+   * Adds as many of the half tier mods as it can as per the order of the filters stat values.
+   * Will show a notification if some mods couldn't be slotted.
+   */
+  const onQuickAddHalfTierMods = () => {
```
Comment (bhollis) [inline src/app/loadout-builder/generated-sets/GeneratedSetButtons.tsx:63]: May have missed the point of the reducer a bit - you're supposed to just send an action like "lock the half tier mods!" and then it calculates the new state given that input. The idea is to decouple the action from the state, or to "dispatch" a function on the state.

**Review by bhollis — APPROVED** [review]
A lot going on here, but I'm willing to try it. I'm guessing from the code that this implements a button that adds all +5 mods that could get you an extra tier - I think that's probably the right flow, and people can delete any mods that get added that they don't want.

File: src/app/loadout-builder/LoadoutBuilder.tsx [hunk 5]
```diff
@@ -95,7 +108,70 @@ function mapStateToProps() {
...
+      let generalSocket: DimSocket | undefined;
+
+      // Look for the first socket in items which has the general category in its whitelist and then break out.
+      itemLoop: for (const item of allItems) {
+        if (item.bucket.inArmor && item.energy) {
+          for (const socket of item.sockets?.allSockets || []) {
+            const socketType = defs.SocketType.get(socket.socketDefinition.socketTypeHash);
+            if (
+              socketType.plugWhitelist.some(
+                (plug) => plug.categoryHash === armor2PlugCategoryHashesByName.general
```
[the remark this answers is not in the document]
Comment (ryan-rushton) [reply src/app/loadout-builder/LoadoutBuilder.tsx:127]: Yeah I was torn about doing it dynamically and just sticking it in known hashes. I might just swap over to known hashes approach.

File: src/app/loadout-builder/LoadoutBuilder.tsx [hunk 6]
```diff
@@ -95,7 +108,70 @@ function mapStateToProps() {
     (loStatSortOrder: number[]) => loStatSortOrder.map((hash) => statHashToType[hash])
   );
 
-  return (state: RootState): StoreProps => {
+  /** A selector to pull out all half tier general mods so we can quick add them to sets. */
+  const halfTierModsSelector = createSelector(
+    allItemsSelector,
+    (state: RootState) => settingsSelector(state).loStatSortOrder,
+    (_: RootState, { defs }: ProvidedProps) => defs,
```
[the remark this answers is not in the document]
Comment (ryan-rushton) [reply src/app/loadout-builder/LoadoutBuilder.tsx:115]: Done

File: src/app/loadout-builder/filter/ModPicker.tsx [hunk 7]
```diff
@@ -155,7 +156,9 @@ function ModPicker({
   onClose,
 }: Props) {
   const [query, setQuery] = useState(initialQuery || '');
-  const [lockedArmor2ModsInternal, setLockedModsInternal] = useState(copy(lockedArmor2Mods));
+  const [lockedArmor2ModsInternal, setLockedModsInternal] = useState(
+    _.mapValues(lockedArmor2Mods, (mods) => mods?.map((mod) => mod.modDef))
```
[the remark this answers is not in the document]
Comment (ryan-rushton) [reply src/app/loadout-builder/filter/ModPicker.tsx:160]: Done

File: src/app/loadout-builder/filter/ModPicker.tsx [hunk 8]
```diff
@@ -214,12 +219,12 @@ function ModPicker({
     return query.length
       ? mods.filter(
           (mod) =>
-            regexp.test(mod.modDef.displayProperties.name) ||
-            regexp.test(mod.modDef.displayProperties.description) ||
-            regexp.test(mod.modDef.itemTypeDisplayName) ||
+            regexp.test(mod.displayProperties.name) ||
```
[the remark this answers is not in the document]
Comment (ryan-rushton) [reply src/app/loadout-builder/filter/ModPicker.tsx:222]: Agreed, I have kinda put it into the 2.0 bucket atm

File: src/app/loadout-builder/generated-sets/GeneratedSetButtons.tsx (hunk 4 above)
Comment (ryan-rushton) [reply src/app/loadout-builder/generated-sets/GeneratedSetButtons.tsx:63]: Yeah just a result of doing things late at night. There is some stuff that will have to stay in this function as the mods are dependent on the armour set but I can move a bunch of it into the reducer. Mainly the stuff around only adding mods that we have space left for.

Comment (ryan-rushton) [reply src/app/loadout-builder/generated-sets/GeneratedSetButtons.tsx:59]: I moved the show notification bit into the reducer as well...I dont know how I feel about that but the notification comes as a side effect of not having enough space for all the sent in mods. 

I could figure this out in the button handler but I pushed it all there for now. Let me know you thoughts and I am happy to move it back to the ui.

**Review by bhollis — APPROVED** [review]

**Pull request merged**